src/parser/parser.ts

```diff
@@ -115,7 +118,22 @@ export function parseAt(source: string, num: number) {
 export function parse(source: string, context: Context) {
   let program: es.Program | undefined
   try {
-    program = acornParse(source, createAcornParserOptions(context)) as unknown as es.Program
+    if (context.variant === Variant.TYPED) {
+      // For Source Typed variant, the code is parsed twice, first using the custom TypeParser and then using Babel Parser.
+      // This is a workaround as the custom TypeParser does not cover all type annotation cases needed for Source Typed
+      // and the Babel Parser does not allow for no semicolon/trailing comma errors when parsing.
+      program = TypeParser.parse(source, createAcornParserOptions(context)) as unknown as es.Program
```

Comment:
`acorn.Node` has its `type` attribute being typed as `string`, so it can refer to any node in an ESTree AST. However, since we know that the AST returned by `acorn.parse` should always start with type `program`, we cast it to `es.Program` from the `estree` library so as to make use of its much more detailed typing.